Assets/HoloToolkit/Input/Scripts/Utilities/Managers/InputManager.cs

```diff
@@ -61,6 +61,146 @@ public class InputManager : Singleton<InputManager>
+            }
+        }
+
+        private void RemoveInputSourceType (uint inputSourceId) {
+            if (inputSourceTypes.ContainsKey(inputSourceId)) {
+                inputSourceTypes.Remove(inputSourceId);
+            }
+        }
+
+        /// <summary>
+        /// Applies a string to the device associated with inputSourceId
+        /// This string will accompany events raised by this device until RemoveEventOrigin is called
+        /// </summary>
+        /// <param name="inputSourceId"></param>
+        /// <param name="eventOrigin"></param>
+        public void ApplyEventOrigin(uint inputSourceId, string eventOrigin) {
+            if (!userEventOriginDictionary.ContainsKey(inputSourceId)) {
+                userEventOriginDictionary.Add(inputSourceId, eventOrigin);
+            } else {
+                userEventOriginDictionary[inputSourceId] = eventOrigin;
+            }
+        }
+
+        /// <summary>
+        /// Removes an event origin string applied via ApplyEventOrigin
+        /// If leaveIfOriginIsDifferent is true, event origin string will not be removed unless it matches the original string
+        /// (This is to prevent pointers from removing tags that don't belong to them)
+        /// </summary>
+        /// <param name="inputSourceId"></param>
+        /// <param name="eventOrigin"></param>
+        /// <param name="leaveIfOriginIsDifferent"></param>
+        public void RemoveEventOrigin(uint inputSourceId, string eventOrigin, bool leaveIfOriginIsDifferent = true) {
+            if (userEventOriginDictionary.ContainsKey(inputSourceId)) {
+                if (!leaveIfOriginIsDifferent || userEventOriginDictionary[inputSourceId] == eventOrigin) {
+                    userEventOriginDictionary.Remove(inputSourceId);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns the event origin associated with this input source, or the default if none is found
+        /// </summary>
+        /// <param name="inputSourceId"></param>
+        /// <returns></returns>
+        private string GetEventOrigin(uint inputSourceId) {
+            if (userEventOriginDictionary.ContainsKey(inputSourceId)) {
+                return userEventOriginDictionary[inputSourceId];
+            } else {
+                return defaultEventOrigin;
+            }
+        }
+
+        /// <summary>
+        /// Source types are defined each time a new device is found
+        /// </summary>
+        private Dictionary<uint, MixedRealityInputType> inputSourceTypes = new Dictionary<uint, MixedRealityInputType>();
+
+        /// <summary>
+        /// User-defined strings set in ApplyEventOrigin
+        /// </summary>
+        private Dictionary<uint, string> userEventOriginDictionary = new Dictionary<uint, string>();
+
+        [SerializeField]
+        private string defaultEventOrigin = string.Empty;
+
+        private void ApplyDefaultEventOrigin(InputSourceInfo newSource) {
+            return;
```

Comment:
unfinished?

Reply:
Unnecessary, part of an earlier experiment.